### ISSUE TITLE
How to recover from a current/assigned state combination without a possible state transition path?

### ISSUE THREAD

**Issue by jnehlmeier** [citusdata/pg_auto_failover#774]
Consider the following state we encountered today during testing:

```
   Name |  Node |    Host:Port |        LSN |   Connection |       Current State |      Assigned State
--------+-------+--------------+------------+--------------+---------------------+--------------------
db002-2 |     4 | db002-2:5432 | 0/6709DFE8 |   read-write |        wait_primary |        wait_primary
db002-1 |     6 | db002-1:5432 | 0/6709C120 | read-write ! |        wait_primary |             demoted



 2021-07-13 07:47:17.440282+02 |    0/6 |             primary |             primary | New state is reported by node 6 "db002-1" (db002-1:5432): "primary"
 
 2021-07-13 11:18:27.359978+02 |    0/6 |             primary |            draining | Setting goal state of node 6 "db002-1" (db002-1:5432) to draining and node 4 "db002-2" (db002-2:5432) to prepare_promotion after a user-initiated failover.
 2021-07-13 11:18:27.359978+02 |    0/4 |           secondary |   prepare_promotion | Setting goal state of node 6 "db002-1" (db002-1:5432) to draining and node 4 "db002-2" (db002-2:5432) to prepare_promotion after a user-initiated failover.
 2021-07-13 11:18:28.320592+02 |    0/6 |             primary |        wait_primary | Setting goal state of node 6 "db002-1" (db002-1:5432) to wait_primary because none of the standby nodes in the quorum are healthy at the moment.
 2021-07-13 11:18:28.503273+02 |    0/6 |        wait_primary |        wait_primary | New state is reported by node 6 "db002-1" (db002-1:5432): "wait_primary"
 2021-07-13 11:18:28.768262+02 |    0/4 |   prepare_promotion |   prepare_promotion | New state is reported by node 4 "db002-2" (db002-2:5432): "prepare_promotion"
 2021-07-13 11:18:28.768262+02 |    0/4 |   prepare_promotion |    stop_replication | Setting goal state of node 6 "db002-1" (db002-1:5432) to demote_timeout and node 4 "db002-2" (db002-2:5432) to stop_replication after node 4 "db002-2" (db002-2:5432) converged to prepare_promotion.
 2021-07-13 11:18:28.768262+02 |    0/6 |        wait_primary |      demote_timeout | Setting goal state of node 6 "db002-1" (db002-1:5432) to demote_timeout and node 4 "db002-2" (db002-2:5432) to stop_replication after node 4 "db002-2" (db002-2:5432) converged to prepare_promotion.
 2021-07-13 11:18:31.832889+02 |    0/4 |    stop_replication |    stop_replication | New state is reported by node 4 "db002-2" (db002-2:5432): "stop_replication"
 2021-07-13 11:23:34.388164+02 |    0/6 |        wait_primary |      demote_timeout | Node node 6 "db002-1" (db002-1:5432) is marked as unhealthy by the monitor
 2021-07-13 11:31:18.036089+02 |    0/4 |    stop_replication |        wait_primary | Setting goal state of node 4 "db002-2" (db002-2:5432) to wait_primary and node 6 "db002-1" (db002-1:5432) to demoted after the demote timeout expired.
 2021-07-13 11:31:18.036089+02 |    0/6 |        wait_primary |             demoted | Setting goal state of node 4 "db002-2" (db002-2:5432) to wait_primary and node 6 "db002-1" (db002-1:5432) to demoted after the demote timeout expired.
 2021-07-13 11:31:19.049409+02 |    0/4 |        wait_primary |        wait_primary | New state is reported by node 4 "db002-2" (db002-2:5432): "wait_primary"
  2021-07-13 11:33:42.51511+02 |    0/4 |        wait_primary |        wait_primary | Node node 4 "db002-2" (db002-2:5432) is marked as unhealthy by the monitor
 2021-07-13 11:35:48.654584+02 |    0/4 |        wait_primary |        wait_primary | Node node 4 "db002-2" (db002-2:5432) is marked as healthy by the monitor
```

Before reaching this state node 6 was primary and node 4 was secondary. We then issued a `pg_autoctl perform failover`  which stopped at 
```
2021-07-13 11:18:31.832889+02 |    0/4 |    stop_replication |    stop_replication | New state is reported by node 4 "db002-2" (db002-2:5432): "stop_replication"
```
for several minutes. Since node 4 stopped replication we figured that node 6 now needs to finally demote but it did not happen. So we restarted node 6 to see if it solves the issue but we ultimately ended up in the above final state. In this state node 6 did not continue anymore because it does not know how to transition from wait_primary to demoted.

We couldn't figure out how to recover from this state. What needs to be done if there is no transition path between current state and assigned state? Do we really have to drop node 6, deleting all data on node 6 and then create a new secondary from it?

We used some pg_autoctl developer/debug commands (standby init, fsm assign/step) in order to convince node 6 to become a secondary but nothing really helped. At the end we fully reset node 6 and its data and registered it again as a new secondary node.

**Comment by DimCitus** [comment]
Hi @jnehlmeier, thanks for another bug report. This one is troubling, having two nodes get both to wait_primary should never happen. Would you have kept pg_autoctl logs from the nodes from that test run?

Other than that, when all things break, the way forward is to drop a node and then create it again. In v1.6.1 it is possible at times to do that without having to delete the PGDATA, and the node can re-join as a secondary from the existing copy of the data set. Sometimes this requires an extra call to `pg_rewind` before hand, and this is not automated in pg_auto_failover yet. Sometimes the needed WAL files might be missing, too, and then you would have to delete the PGDATA and start afresh anyway.

**Comment by jnehlmeier** [comment]
> Hi @jnehlmeier, thanks for another bug report. This one is troubling, having two nodes get both to wait_primary should never happen. Would you have kept pg_autoctl logs from the nodes from that test run?

Below are logs collected during this test. However the three hosts that run the containers did not had synchronized times so logging timestamps are not easily comparable. I have added the diff in seconds to the headline of each log section below.

As you can see in the logs the monitor did not proceed (and during that pause we restarted the nodes multiple times)
```
"11:18:31 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): stop_replication ➜ stop_replication"
"11:30:53 1 INFO  pg_autoctl received signal SIGTERM, terminating"
```
because the monitor has put node 6 already in an unreachable state
```
MONITOR:

"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to draining and node 4 ""db002-2"" (db002-2:5432) to prepare_promotion after a user-initiated failover."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): primary ➜ draining"
"11:18:28 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): secondary ➜ prepare_promotion"
"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to wait_primary because none of the standby nodes in the quorum are healthy at the moment."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): primary ➜ wait_primary"
"11:18:28 14 INFO  New state is reported by node 6 ""db002-1"" (db002-1:5432): ""wait_primary"""
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ wait_primary"
"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to demote_timeout and node 4 ""db002-2"" (db002-2:5432) to stop_replication after node 4 ""db002-2"" (db002-2:5432) converged to prepare_promotion."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ demote_timeout"


NODE 6 (~+11s timestamp diff):

"11:18:39 30 INFO  Monitor assigned new state ""wait_primary"""
"11:18:39 30 INFO  FSM transition from ""primary"" to ""wait_primary"": Secondary became unhealthy"
"11:18:39 30 INFO  Disabling synchronous replication"
"11:18:39 30 INFO  Reloading Postgres configuration and HBA rules"
"11:18:39 30 INFO  Transition complete: current state is now ""wait_primary"""
"11:18:39 30 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ prepare_promotion"
"11:18:39 30 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ stop_replication"
"11:18:39 30 INFO  New state for this node (node 6, ""db002-1"") (db002-1:5432): wait_primary ➜ demote_timeout"
"11:18:39 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:39 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:39 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
```
---
### Full logs


Monitor logs:
```
"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to draining and node 4 ""db002-2"" (db002-2:5432) to prepare_promotion after a user-initiated failover."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): primary ➜ draining"
"11:18:28 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): secondary ➜ prepare_promotion"
"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to wait_primary because none of the standby nodes in the quorum are healthy at the moment."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): primary ➜ wait_primary"
"11:18:28 14 INFO  New state is reported by node 6 ""db002-1"" (db002-1:5432): ""wait_primary"""
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ wait_primary"
"11:18:28 14 INFO  Setting goal state of node 6 ""db002-1"" (db002-1:5432) to demote_timeout and node 4 ""db002-2"" (db002-2:5432) to stop_replication after node 4 ""db002-2"" (db002-2:5432) converged to prepare_promotion."
"11:18:28 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ demote_timeout"
"11:18:28 14 INFO  New state is reported by node 4 ""db002-2"" (db002-2:5432): ""prepare_promotion"""
"11:18:28 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ prepare_promotion"
"11:18:28 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ stop_replication"
"11:18:31 14 INFO  New state is reported by node 4 ""db002-2"" (db002-2:5432): ""stop_replication"""
"11:18:31 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): stop_replication ➜ stop_replication"
"11:30:53 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:30:53 14 INFO  Listener service received signal SIGTERM, terminating"
"11:30:53 13 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:30:53 13 INFO  Stopping pg_autoctl postgres service"
"11:30:53 13 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/monitor --wait stop --mode fast"
"11:30:53 1 INFO  Waiting for subprocesses to terminate."
"11:30:53 1 ERROR Unknown subprocess died with pid 24"
"11:30:53 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass ... "
"Monitor already created."
"Generating pg_hba.conf using allowed subnet 10.41.1.0/24 ..."
"Starting pg_autoctl..."
"11:30:55 1 INFO  Started pg_autoctl postgres service with pid 13"
"11:30:55 13 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/monitor -v"
"11:30:55 1 INFO  Started pg_autoctl listener service with pid 14"
"11:30:55 14 INFO   /usr/bin/pg_autoctl do service listener --pgdata /var/lib/postgresql/monitor -v"
"11:30:55 14 INFO  Managing the monitor at postgres://autoctl_node@db002-monitor:5432/pg_auto_failover?sslmode=require"
"11:30:55 14 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/monitor/pg_autoctl.cfg"""
"11:30:56 23 INFO   /usr/lib/postgresql/13/bin/postgres -D /var/lib/postgresql/monitor -p 5432 -h *"
"11:30:56 14 INFO  The version of extension ""pgautofailover"" is ""1.5"" on the monitor"
"11:30:56 14 INFO  Contacting the monitor to LISTEN to its events."
"11:30:56 13 INFO  Postgres is now serving PGDATA ""/var/lib/postgresql/monitor"" on port 5432 with pid 23"
"11:31:18 14 INFO  Setting goal state of node 4 ""db002-2"" (db002-2:5432) to wait_primary and node 6 ""db002-1"" (db002-1:5432) to demoted after the demote timeout expired."
"11:31:18 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): stop_replication ➜ wait_primary"
"11:31:18 14 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ demoted"
"11:31:19 14 INFO  New state is reported by node 4 ""db002-2"" (db002-2:5432): ""wait_primary"""
"11:31:19 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): wait_primary ➜ wait_primary"
"13:10:25 14 INFO  Node node 4 ""db002-2"" (db002-2:5432) is marked as unhealthy by the monitor"
"13:10:25 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): wait_primary ➜ wait_primary"
"13:10:25 14 INFO  Node node 4 ""db002-2"" (db002-2:5432) is marked as healthy by the monitor"
"13:10:25 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): wait_primary ➜ wait_primary"
"13:10:25 14 INFO  Removing node 6 ""db002-1"" (db002-1:5432) from formation ""default"" and group 0"
"13:10:25 14 INFO  Setting number_sync_standbys to 0 for formation ""default"" now that we have 0 standby nodes set with replication-quorum."
"13:10:25 14 INFO  Setting goal state of node 4 ""db002-2"" (db002-2:5432) to single as there is no other node."
"13:10:25 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): wait_primary ➜ single"
"13:10:25 14 INFO  Setting goal state of node 4 ""db002-2"" (db002-2:5432) to single as there is no other node."
"13:10:25 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): wait_primary ➜ single"
"13:10:26 14 INFO  New state is reported by node 4 ""db002-2"" (db002-2:5432): ""single"""
"13:10:26 14 INFO  New state for node 4 ""db002-2"" (db002-2:5432): single ➜ single"
"13:48:17 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"13:48:17 13 INFO  Postgres controller service received signal SIGTERM, terminating"
"13:48:17 14 INFO  Listener service received signal SIGTERM, terminating"
"13:48:17 13 INFO  Stopping pg_autoctl postgres service"
"13:48:17 13 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/monitor --wait stop --mode fast"
"13:48:17 1 ERROR Unknown subprocess died with pid 24"
"13:48:18 1 INFO  Stop pg_autoctl"
```
---

Node 6 (was `primary` before `pg_autoctl perform failover`, timestamps are ~+11s to monitor) 
```
"11:18:39 30 INFO  Monitor assigned new state ""wait_primary"""
"11:18:39 30 INFO  FSM transition from ""primary"" to ""wait_primary"": Secondary became unhealthy"
"11:18:39 30 INFO  Disabling synchronous replication"
"11:18:39 30 INFO  Reloading Postgres configuration and HBA rules"
"11:18:39 30 INFO  Transition complete: current state is now ""wait_primary"""
"11:18:39 30 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ prepare_promotion"
"11:18:39 30 INFO  New state for node 4 ""db002-2"" (db002-2:5432): prepare_promotion ➜ stop_replication"
"11:18:39 30 INFO  New state for this node (node 6, ""db002-1"") (db002-1:5432): wait_primary ➜ demote_timeout"
"11:18:39 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:39 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:39 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:18:40 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:40 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:40 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:18:41 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:41 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:41 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:18:42 30 INFO  New state for node 4 ""db002-2"" (db002-2:5432): stop_replication ➜ stop_replication"
"11:18:42 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:42 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:42 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:18:43 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:18:43 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:18:43 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "

... LOTS OF RETRIES ...

"11:23:33 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:23:33 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:23:33 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:23:34 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:23:34 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:23:34 29 INFO  Stopping pg_autoctl postgres service"
"11:23:34 29 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/node --wait stop --mode fast"
"11:23:34 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:23:34 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:23:34 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:23:34 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:23:35 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:23:35 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:23:35 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:23:35 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:23:35 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:23:35 30 INFO  pg_autoctl service is running, current state is ""wait_primary"""
"11:23:35 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:23:36 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:23:36 30 INFO  Checking for HBA rules for node 4 ""db002-2"" (db002-2:5432)"
"11:23:36 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-2 trust"
"11:23:36 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-2 trust"
"11:23:36 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:23:36 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:23:36 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:23:37 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:23:37 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:23:37 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:23:37 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "

... LOTS OF RETRIES ...

"11:29:56 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:29:56 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:29:56 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:29:56 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:29:57 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:29:57 30 INFO  Monitor assigned new state ""demote_timeout"""
"11:29:57 30 FATAL pg_autoctl does not know how to reach state ""demote_timeout"" from ""wait_primary"""
"11:29:57 30 ERROR Failed to transition to state ""demote_timeout"", retrying... "
"11:29:58 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:29:58 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:29:58 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:32:25 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:32:25 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:32:25 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:32:25 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:32:25 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:32:25 30 INFO  pg_autoctl service is running, current state is ""wait_primary"""
"11:32:25 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:32:25 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:32:25 30 INFO  Checking for HBA rules for node 4 ""db002-2"" (db002-2:5432)"
"11:32:25 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-2 trust"
"11:32:25 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-2 trust"
"11:32:25 30 INFO  Monitor assigned new state ""demoted"""
"11:32:25 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:32:25 30 ERROR Failed to transition to state ""demoted"", retrying... "
"11:32:27 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:32:27 30 INFO  Monitor assigned new state ""demoted"""
"11:32:27 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:32:27 30 ERROR Failed to transition to state ""demoted"", retrying... "

... LOTS OF RETRIES ....

"11:33:32 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:33:33 30 INFO  Monitor assigned new state ""demoted"""
"11:33:33 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:33:33 30 ERROR Failed to transition to state ""demoted"", retrying... "
"11:33:34 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:33:34 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:33:34 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:33:34 30 INFO  Monitor assigned new state ""demoted"""
"11:33:34 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:33:34 30 ERROR Failed to transition to state ""demoted"", retrying... "
"11:33:34 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ....



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:33:52 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:33:52 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:33:52 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:33:52 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:33:52 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:33:52 30 INFO  pg_autoctl service is running, current state is ""wait_primary"""
"11:33:52 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:33:53 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:33:53 30 INFO  Checking for HBA rules for node 4 ""db002-2"" (db002-2:5432)"
"11:33:53 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-2 trust"
"11:33:53 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-2 trust"
"11:33:53 30 INFO  Monitor assigned new state ""demoted"""
"11:33:53 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:33:53 30 ERROR Failed to transition to state ""demoted"", retrying... "
"11:33:54 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:33:54 30 INFO  Monitor assigned new state ""demoted"""
"11:33:54 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:33:54 30 ERROR Failed to transition to state ""demoted"", retrying... "
"11:33:55 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:33:55 30 INFO  Monitor assigned new state ""demoted"""
"11:33:55 30 FATAL pg_autoctl does not know how to reach state ""demoted"" from ""wait_primary"""
"11:33:55 30 ERROR Failed to transition to state ""demoted"", retrying... "

... ENDLESS RETRIES UNTIL WE STOPPED THE TEST ...
```
---
Node 4 (was `secondary` before `pg_autoctl perform failover`, timestamps are ~+13s to monitor)
```
"11:18:41 30 ERROR Monitor ERROR:  deadlock detected"
"11:18:41 30 ERROR Monitor DETAIL:  Process 616799 waits for ExclusiveLock on advisory lock [16385,822708183,0,11]; blocked by process 616803."
"11:18:41 30 ERROR Monitor Process 616803 waits for ShareLock on transaction 341083; blocked by process 616799."
"11:18:41 30 ERROR Monitor HINT:  See server log for query details."
"11:18:41 30 ERROR Failed to get node state for node 4 in group 0 of formation ""default"" with initial state ""secondary"", replication state """", and current lsn ""0/6709C120"", see previous lines for details"
"11:18:41 30 ERROR Failed to get the goal state from the monitor"
"11:18:41 30 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ wait_primary"
"11:18:41 30 INFO  Successfully got the goal state from the monitor"
"11:18:41 30 INFO  FSM transition from ""secondary"" to ""prepare_promotion"": Stop traffic to primary, wait for it to finish draining."
"11:18:41 30 INFO  Monitor assigned new state ""prepare_promotion"""
"11:18:41 30 INFO  Transition complete: current state is now ""prepare_promotion"""
"11:18:42 30 INFO  FSM transition from ""prepare_promotion"" to ""stop_replication"": Prevent against split-brain situations."
"11:18:42 30 INFO  Setting default_transaction_read_only to on"
"11:18:42 30 INFO  Monitor assigned new state ""stop_replication"""
"11:18:42 30 INFO  Prevent writes to the promoted standby while the primary is not demoted yet, by making the service incompatible with target_session_attrs = read-write"
"11:18:42 30 INFO  Reloading Postgres configuration and HBA rules"
"11:18:42 30 INFO  Promoting postgres"
"11:18:42 30 INFO  Waiting for postgres to promote"
"11:18:44 30 INFO  Cleaning-up Postgres replication settings"
"11:18:44 30 INFO  Disabling synchronous replication"
"11:18:44 30 INFO  Reloading Postgres configuration and HBA rules"
"11:18:45 30 INFO  Transition complete: current state is now ""stop_replication"""
"11:18:45 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): prepare_promotion ➜ prepare_promotion"
"11:18:45 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): prepare_promotion ➜ stop_replication"
"11:18:45 30 INFO  New state for node 6 ""db002-1"" (db002-1:5432): wait_primary ➜ demote_timeout"
"11:18:45 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): stop_replication ➜ stop_replication"
"11:20:57 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:20:57 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:20:57 29 INFO  Stopping pg_autoctl postgres service"
"11:20:57 29 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/node --wait stop --mode fast"
"11:20:57 1 INFO  Waiting for subprocesses to terminate."
"11:20:59 1 ERROR Unknown subprocess died with pid 105884"
"11:20:59 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:21:02 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:21:02 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:21:02 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:21:02 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:21:02 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:21:02 30 INFO  Contents of ""/var/lib/postgresql/node/postgresql-auto-failover.conf"" have changed, overwriting"
"11:21:02 30 INFO  pg_autoctl service is running, current state is ""stop_replication"""
"11:21:02 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:21:02 30 INFO  Checking for HBA rules for node 6 ""db002-1"" (db002-1:5432)"
"11:21:02 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-1 trust"
"11:21:02 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-1 trust"
"11:21:02 38 INFO   /usr/lib/postgresql/13/bin/postgres -D /var/lib/postgresql/node -p 5432 -h *"
"11:21:03 29 INFO  Postgres is now serving PGDATA ""/var/lib/postgresql/node"" on port 5432 with pid 38"
"11:21:03 30 WARN  PostgreSQL was not running, restarted with pid 38"
"11:30:06 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"11:30:06 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"11:30:06 29 INFO  Stopping pg_autoctl postgres service"
"11:30:06 29 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/node --wait stop --mode fast"
"11:30:06 1 INFO  Waiting for subprocesses to terminate."
"11:30:06 1 ERROR Unknown subprocess died with pid 39"
"11:30:07 1 INFO  Stop pg_autoctl"



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:31:31 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:31:31 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:31:31 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:31:31 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:31:31 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:31:31 30 INFO  pg_autoctl service is running, current state is ""stop_replication"""
"11:31:31 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:31:31 30 INFO  Checking for HBA rules for node 6 ""db002-1"" (db002-1:5432)"
"11:31:31 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-1 trust"
"11:31:31 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-1 trust"
"11:31:31 30 INFO  Monitor assigned new state ""wait_primary"""
"11:31:31 38 INFO   /usr/lib/postgresql/13/bin/postgres -D /var/lib/postgresql/node -p 5432 -h *"
"11:31:31 29 INFO  Postgres is now serving PGDATA ""/var/lib/postgresql/node"" on port 5432 with pid 38"
"11:31:32 30 WARN  PostgreSQL was not running, restarted with pid 38"
"11:31:32 30 INFO  FSM transition from ""stop_replication"" to ""wait_primary"": Confirmed promotion with the monitor"
"11:31:32 30 INFO  Setting default_transaction_read_only to off"
"11:31:32 30 INFO  Reloading Postgres configuration and HBA rules"
"11:31:32 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:31:32 30 INFO  Checking for HBA rules for node 6 ""db002-1"" (db002-1:5432)"
"11:31:32 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-1 trust"
"11:31:32 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-1 trust"
"11:31:32 30 INFO  Transition complete: current state is now ""wait_primary"""

... SOME SHUTDOWN LOGS MISSING ...



... USER-INITIATED CONTAINER RESTART ...



"Given log level not defined or unknown"
"Using pg_auto_failover log level INFO ..."
"Generating .pgpass file ... "
"pg_auto_failover node already created."
"Setting replication.password ..."
"Node configuration done."
"Starting pg_autoctl..."
"11:35:59 1 INFO  Started pg_autoctl postgres service with pid 29"
"11:35:59 29 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v"
"11:35:59 1 INFO  Started pg_autoctl node-active service with pid 30"
"11:35:59 30 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v"
"11:35:59 30 INFO  Reloaded the new configuration from ""/home/postgres/.config/pg_autoctl/var/lib/postgresql/node/pg_autoctl.cfg"""
"11:35:59 30 INFO  pg_autoctl service is running, current state is ""wait_primary"""
"11:35:59 30 WARN  Failed to update the keeper's state from the local PostgreSQL instance."
"11:35:59 30 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes."
"11:35:59 30 INFO  Checking for HBA rules for node 6 ""db002-1"" (db002-1:5432)"
"11:35:59 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication ""pgautofailover_replicator"" db002-1 trust"
"11:35:59 30 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl ""postgres"" ""pgautofailover_replicator"" db002-1 trust"
"11:35:59 38 INFO   /usr/lib/postgresql/13/bin/postgres -D /var/lib/postgresql/node -p 5432 -h *"
"11:35:59 30 WARN  PostgreSQL was not running, restarted with pid 38"
"11:35:59 29 INFO  Postgres is now serving PGDATA ""/var/lib/postgresql/node"" on port 5432 with pid 38"
"11:36:00 30 INFO  Updated the keeper's state from the local PostgreSQL instance, which is running"
"11:36:00 30 INFO  pg_autoctl managed to ensure current state ""wait_primary"": PostgreSQL is running"
"12:11:37 30 INFO  Creating replication slot ""pgautofailover_standby_6"""
"13:10:39 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): wait_primary ➜ single"
"13:10:39 30 INFO  Monitor assigned new state ""single"""
"13:10:39 30 INFO  Dropping replication slot ""pgautofailover_standby_6"""
"13:10:39 30 INFO  FSM transition from ""wait_primary"" to ""single"": Other node was forcibly removed, now single"
"13:10:39 30 INFO  Disabling synchronous replication"
"13:10:39 30 INFO  Reloading Postgres configuration and HBA rules"
"13:10:39 30 INFO  Transition complete: current state is now ""single"""
"13:10:39 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): wait_primary ➜ single"
"13:10:39 30 INFO  New state for this node (node 4, ""db002-2"") (db002-2:5432): single ➜ single"
"13:34:40 1 INFO  pg_autoctl received signal SIGTERM, terminating"
"13:34:40 29 INFO  Postgres controller service received signal SIGTERM, terminating"
"13:34:40 29 INFO  Stopping pg_autoctl postgres service"
"13:34:40 29 INFO  /usr/lib/postgresql/13/bin/pg_ctl --pgdata /var/lib/postgresql/node --wait stop --mode fast"
"13:34:40 1 INFO  Waiting for subprocesses to terminate."
"13:34:40 1 ERROR Unknown subprocess died with pid 39"
"13:34:40 1 INFO  Stop pg_autoctl"
```

**Comment by DimCitus** [comment]
On a first reading it looks like the secondary node is failing right after a manual failover has been started. Is that correct? If that's the case, I will play with testing this scenario and adding coverage in the test suite (unless we already have it).

Also please note that if we're seeing a failure during a failover operation, then we are in the territory of “multiple node failures” which where we have yet work to do. Our model at the moment is meant to fully tolerate the failure of any one node at a time, and knows how to handle multiple failures in some cases only.

**Comment by jnehlmeier** [comment]
Both postgres nodes were running at the time of failover. I believe there was some docker networking hiccup and pg_auto_failover already knew that the secondary is gone but did not take any action yet (maybe some timeout wasn't reached yet). However it did take the action as soon as we manually triggered the failover process. Sadly it does it at a bad time and breaks FSM.

Basically the primary should progress primary -> draining -> demote_timeout. But during that transition the monitor sets the state of the primary to wait_primary since the secondary wasn't reachable. When the monitor does this, it should not continue with the usual (requested) failover transition because that's why we end up in wait_primary -> demote_timeout.

**Comment by dimitri** [comment]
Fixed by #1165 (merged).

**Root cause:** a race between two independent rules in the monitor's `group_state_machine.c`. When a failover candidate converges out of `secondary` (e.g. to `prepare_promotion`), the count of healthy standbys in the quorum drops to zero for that tick — not because a standby was actually lost, but because the candidate itself is mid-promotion. A separate rule reads that zero count as "no healthy standby in the quorum" and reassigns the primary's goal state back to `wait_primary`, in the same window another rule can assign it `demote_timeout`. There is no edge from `wait_primary` to `demote_timeout` in the keeper's state machine, so once the primary landed on `wait_primary` with a `demote_timeout` goal, it had no legal transition to make and just kept retrying (logged as `FATAL`, though that's a severity label here, not a crash — the keeper loops and retries rather than exiting). That race, hit during your `pg_autoctl perform failover`, is exactly what produced `db002-1` stuck reporting `wait_primary` against goal `demote_timeout`.

**The fix:** the "no healthy standby in the quorum" rule now checks whether a failover is already in progress before reassigning the primary to `wait_primary`, so it no longer fires on a candidate's own promotion in flight. A new pg_regress test (`failover_candidate_leaves_secondary`) manufactures this exact scenario directly and asserts the primary's goal state is no longer disturbed while its candidate is promoting.

The pre-existing drain-timeout mechanism (added in #707, already on main when this was filed) that eventually moves a stuck `wait_primary` primary to `demoted` is unrelated and untouched — it's what got `db002-1` unstuck in your log after ~13 minutes, but only after the race had already put it in a state it couldn't otherwise resolve on its own. With this fix, the race itself doesn't happen, so a primary should no longer land in an unreachable goal state during an ordinary failover.

**Comment by jnehlmeier** [comment]
@dimitri Interesting, just recently we had that situation again while a switch went rogue and network was unstable.

During that unstable network a different cluster had settled on a state where one node was `primary` and the other was `wait_primary`. The `wait_primary` did accept writes and the `primary` not. 

Basically the monitor did a failover after the `primary` became unhealthy. It managed to stop replication and assigned the secondary the new state `wait_primary`. During that process it lost connection to the previous `primary` which then came back some moments later and causes the monitor to assign `catchingup` to the previous `primary`. As there is no edge from `primary` to `catchingup` we had remove it from the cluster and re-init it again.

So maybe there are other races if that is not already covered by the fix you did.


Here is the monitor log if you are interested

```
2026-07-02T16:58:42.250204147Z 18:58:42 15 INFO  Node node 1 "db001-1" (db001-1:5432) is marked as unhealthy by the monitor
2026-07-02T16:58:42.250225269Z 18:58:42 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ primary
2026-07-02T16:58:42.263543924Z 18:58:42 15 INFO  Node node 1 "db001-1" (db001-1:5432) is marked as healthy by the monitor
2026-07-02T16:58:42.263562316Z 18:58:42 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ primary
2026-07-02T17:01:55.248480049Z 19:01:55 15 INFO  Setting goal state of node 1 "db001-1" (db001-1:5432) to draining and node 151 "db001-2" (db001-2:5432) to prepare_promotion after node 1 "db001-1" (db001-1:5432) became unhealthy.
2026-07-02T17:01:55.248483416Z 19:01:55 15 INFO  New state for node 151 "db001-2" (db001-2:5432): secondary ➜ prepare_promotion
2026-07-02T17:01:55.248483989Z 19:01:55 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ draining
2026-07-02T17:01:55.374151171Z 19:01:55 15 INFO  New state is reported by node 151 "db001-2" (db001-2:5432): "prepare_promotion"
2026-07-02T17:01:55.374155405Z 19:01:55 15 INFO  New state for node 151 "db001-2" (db001-2:5432): prepare_promotion ➜ prepare_promotion
2026-07-02T17:01:55.374157374Z 19:01:55 15 INFO  Setting goal state of node 1 "db001-1" (db001-1:5432) to demote_timeout and node 151 "db001-2" (db001-2:5432) to stop_replication after node 151 "db001-2" (db001-2:5432) converged to prepare_promotion.
2026-07-02T17:01:55.374160303Z 19:01:55 15 INFO  New state for node 151 "db001-2" (db001-2:5432): prepare_promotion ➜ stop_replication
2026-07-02T17:01:55.374585263Z 19:01:55 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ demote_timeout
2026-07-02T17:02:06.297436352Z 19:02:06 15 INFO  New state is reported by node 1 "db001-1" (db001-1:5432): "demote_timeout"
2026-07-02T17:02:06.297451908Z 19:02:06 15 INFO  New state for node 1 "db001-1" (db001-1:5432): demote_timeout ➜ demote_timeout
2026-07-02T17:02:14.799722890Z 19:02:14 15 INFO  New state is reported by node 151 "db001-2" (db001-2:5432): "stop_replication"
2026-07-02T17:02:14.799747524Z 19:02:14 15 INFO  New state for node 151 "db001-2" (db001-2:5432): stop_replication ➜ stop_replication
2026-07-02T17:02:14.799749079Z 19:02:14 15 INFO  Setting goal state of node 151 "db001-2" (db001-2:5432) to wait_primary and node 1 "db001-1" (db001-1:5432) to demoted after the demote timeout expired.
2026-07-02T17:02:14.799750025Z 19:02:14 15 INFO  New state for node 151 "db001-2" (db001-2:5432): stop_replication ➜ wait_primary
2026-07-02T17:02:14.799750827Z 19:02:14 15 INFO  New state for node 1 "db001-1" (db001-1:5432): demote_timeout ➜ demoted
2026-07-02T17:02:14.896894910Z 19:02:14 15 INFO  New state is reported by node 151 "db001-2" (db001-2:5432): "wait_primary"
2026-07-02T17:02:14.897052366Z 19:02:14 15 INFO  New state for node 151 "db001-2" (db001-2:5432): wait_primary ➜ wait_primary
2026-07-02T17:02:14.901331057Z 19:02:14 15 INFO  New state is reported by node 1 "db001-1" (db001-1:5432): "demoted"
2026-07-02T17:02:14.901332778Z 19:02:14 15 INFO  New state for node 1 "db001-1" (db001-1:5432): demoted ➜ demoted
2026-07-02T17:02:14.901333111Z 19:02:14 15 INFO  Setting goal state of node 1 "db001-1" (db001-1:5432) to catchingup after it converged to demotion and node 151 "db001-2" (db001-2:5432) converged to wait_primary.
2026-07-02T17:02:14.901333332Z 19:02:14 15 INFO  New state for node 1 "db001-1" (db001-1:5432): demoted ➜ catchingup

// here old primary comes back and reports its primary state causing the primary -> catchingup edge

2026-07-02T17:02:36.713150323Z 19:02:36 15 INFO  New state is reported by node 1 "db001-1" (db001-1:5432): "primary"
2026-07-02T17:02:36.713160019Z 19:02:36 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup
2026-07-02T17:02:44.495179189Z 19:02:44 15 INFO  Node node 1 "db001-1" (db001-1:5432) is marked as unhealthy by the monitor
2026-07-02T17:02:44.495181356Z 19:02:44 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup
2026-07-02T17:10:07.699003091Z 19:10:07 15 INFO  Node node 1 "db001-1" (db001-1:5432) is marked as healthy by the monitor
2026-07-02T17:10:07.699013334Z 19:10:07 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup
2026-07-02T17:29:45.195389305Z 19:29:45 1 ERROR Unknown subprocess died with pid 1910491
2026-07-02T17:54:35.619450836Z 19:54:35 15 INFO  Node node 1 "db001-1" (db001-1:5432) is marked as unhealthy by the monitor
2026-07-02T17:54:35.619482724Z 19:54:35 15 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup

// here we removed it

2026-07-03T10:23:43.677185143Z 12:23:43 15 INFO  Removing node 1 "db001-1" (db001-1:5432) from formation "default" and group 0

```

**Comment by dimitri** [comment]
Dug into this one against current `main`.

**The stuck state itself:** the monitor accepts whatever state a node self-reports with no check that it's a legal follow-on from what it already knows (`NodeActive()` in `node_active_protocol.c` just overwrites `reportedState`). There's already a targeted guard for a node unexpectedly reporting `demote_timeout` with an unreachable goal (added for #1025), but none for a node reporting `primary` while its goal is `catchingup` -- so once that report lands, nothing re-targets it and it's stuck until manually removed. That part is real and fixable.

**The "demoted -> primary" flip itself, though, isn't explainable from the monitor's log alone.** There's no transition back to `primary` from `demoted` in the keeper's FSM (`fsm.c`). And since node1 had already successfully reported `demote_timeout` then `demoted` -- meaning its keeper was actively cooperating with the failover, not partitioned -- the self-fence timeout logic doesn't apply here either. Beyond that, we're guessing (a stale/orphaned second process is the most plausible shape, but that's not something the monitor's side can confirm).

@jnehlmeier: do you still have node1's own `pg_autoctl` logs (not the monitor's) from that window? Specifically useful: whether there's a second process start (a new PID) around 19:02, or any sign of an old process still running. Without that, there's nothing more to conclude from the monitor's report alone.

**Comment by jnehlmeier** [comment]
@dimitri We run everything in docker swarm and `pg_autoctl` does not write a log file, so I don't have the concrete `pg_autoctl` logs for that node for that time anymore.

However I can see that a new container for `db001-1 (node 1)` had been started at `19:01:22` and terminated at `19:01:47` because it could not connect to the monitor and gave up after 5 times within 20 seconds.

```
2026-07-02T17:01:22.255245737Z Starting pg_autoctl...
2026-07-02T17:01:22.272330523Z 19:01:22 1 INFO  Started pg_autoctl postgres service with pid 30
2026-07-02T17:01:22.272402133Z 19:01:22 30 INFO   /usr/bin/pg_autoctl do service postgres --pgdata /var/lib/postgresql/node -v
2026-07-02T17:01:22.272551306Z 19:01:22 1 INFO  Started pg_autoctl node-active service with pid 31
2026-07-02T17:01:22.272674040Z 19:01:22 31 INFO   /usr/bin/pg_autoctl do service node-active --pgdata /var/lib/postgresql/node -v
2026-07-02T17:01:26.288627422Z 19:01:26 31 ERROR Connection to database failed: connection to server at "db001-monitor" (10.41.1.75), port 5432 failed: timeout expired
2026-07-02T17:01:26.288636698Z 19:01:26 31 ERROR connection to server at "db001-monitor" (10.41.1.193), port 5432 failed: timeout expired
2026-07-02T17:01:26.288637052Z 19:01:26 31 ERROR Failed to connect to monitor database at "postgres://autoctl_node@db001-monitor:5432/pg_auto_failover?sslmode=require", see above for details
2026-07-02T17:01:26.288676266Z 19:01:26 31 ERROR Failed to get nodes for group 0 in formation "default" from the monitor
2026-07-02T17:01:26.288678694Z 19:01:26 31 ERROR Failed to query monitor to see if node id 1 has been dropped already
2026-07-02T17:01:26.378916001Z 19:01:26 1 WARN  pg_autoctl service node-active exited with exit status 0
2026-07-02T17:01:26.378918622Z 19:01:26 1 INFO  Restarting service node-active

// ... repeat 4 times until shutdown
```

I know that docker swarm also had trouble because of the unstable network and it could every well be that it had tried to launch the db node again.

What I have is the postgres logs of `db001-1 (node 1)` and it does not look like as if two processes log into that file in parallel (two containers would have the same volume mounted and work in the same directory). However the log contains a `promote request`.

```
// Many 4-lines blocks with restartpoints...

2026-07-02 18:53:19 CEST  DETAIL:  Last completed transaction was at log time 2026-07-02 18:53:18.397982+02.
2026-07-02 18:56:10 CEST  LOG:  restartpoint starting: time
2026-07-02 18:59:13 CEST  LOG:  restartpoint complete: wrote 1827 buffers (0.3%); 0 WAL file(s) added, 0 removed, 1 recycled; write=182.791 s, sync=0.003 s, total=182.849 s; sync files=116, longest=0.003 s, average=0.001 s; distance=13696 kB, estimate=26687 kB
2026-07-02 18:59:13 CEST  LOG:  recovery restart point at 9B0/B549AE30
2026-07-02 18:59:13 CEST  DETAIL:  Last completed transaction was at log time 2026-07-02 18:59:12.191363+02.
2026-07-02 19:01:10 CEST  LOG:  restartpoint starting: time
2026-07-02 19:01:20 CEST  LOG:  restartpoint complete: wrote 98 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=9.791 s, sync=0.003 s, total=9.829 s; sync files=51, longest=0.003 s, average=0.001 s; distance=13335 kB, estimate=25351 kB
2026-07-02 19:01:20 CEST  LOG:  recovery restart point at 9B0/B61A0C70
2026-07-02 19:01:20 CEST  DETAIL:  Last completed transaction was at log time 2026-07-02 19:01:18.48646+02.

// SIGHUP 

2026-07-02 19:01:55 CEST  LOG:  received SIGHUP, reloading configuration files
2026-07-02 19:01:55 CEST  LOG:  parameter "default_transaction_read_only" changed to "on"

// promote request comes in

2026-07-02 19:01:55 CEST  LOG:  received promote request
2026-07-02 19:01:55 CEST  FATAL:  terminating walreceiver process due to administrator command
2026-07-02 19:01:55 CEST  LOG:  invalid resource manager ID 46 at 9B0/B6313298
2026-07-02 19:01:55 CEST  LOG:  redo done at 9B0/B6313270 system usage: CPU: user: 62719.53 s, system: 29282.33 s, elapsed: 48406961.86 s
2026-07-02 19:01:55 CEST  LOG:  last completed transaction was at log time 2026-07-02 19:01:55.513226+02
2026-07-02 19:02:13 CEST  LOG:  selected new timeline ID: 20
2026-07-02 19:02:13 CEST  LOG:  archive recovery complete
2026-07-02 19:02:13 CEST  LOG:  checkpoint starting: force
2026-07-02 19:02:14 CEST  LOG:  database system is ready to accept connections
2026-07-02 19:02:14 CEST  LOG:  checkpoint complete: wrote 121 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.728 s, sync=0.003 s, total=0.763 s; sync files=39, longest=0.003 s, average=0.001 s; distance=1481 kB, estimate=22964 kB
2026-07-02 19:02:14 CEST  LOG:  checkpoint starting: immediate force wait
2026-07-02 19:02:14 CEST  LOG:  checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.010 s, sync=0.001 s, total=0.020 s; sync files=0, longest=0.000 s, average=0.000 s; distance=0 kB, estimate=20668 kB

// SIGHUB

2026-07-02 19:02:14 CEST  LOG:  received SIGHUP, reloading configuration files
2026-07-02 19:02:14 CEST  LOG:  parameter "primary_conninfo" removed from configuration file, reset to default
2026-07-02 19:02:14 CEST  LOG:  parameter "primary_slot_name" removed from configuration file, reset to default
2026-07-02 19:02:14 CEST  LOG:  parameter "recovery_target_timeline" cannot be changed without restarting the server
2026-07-02 19:02:14 CEST  LOG:  configuration file "/var/lib/postgresql/node/postgresql.conf" contains errors; unaffected changes were applied

// SIGHUB

2026-07-02 19:02:14 CEST  LOG:  received SIGHUP, reloading configuration files
2026-07-02 19:02:14 CEST  LOG:  parameter "recovery_target_timeline" cannot be changed without restarting the server
2026-07-02 19:02:14 CEST  LOG:  parameter "default_transaction_read_only" changed to "off"
2026-07-02 19:02:14 CEST  LOG:  configuration file "/var/lib/postgresql/node/postgresql.conf" contains errors; unaffected changes were applied

// slow query logs
```

At the same time the `pg_autoctl` logs of `db001-2` which was promoted:

```
// db001-2: promoting...

2026-07-02T17:01:55.282740630Z 19:01:55 31 INFO  Monitor assigned new state "prepare_promotion"
2026-07-02T17:01:55.283614453Z 19:01:55 31 INFO  FSM transition from "secondary" to "prepare_promotion": Stop traffic to primary, wait for it to finish draining.
2026-07-02T17:01:55.283617413Z 19:01:55 31 INFO  Transition complete: current state is now "prepare_promotion"
2026-07-02T17:01:55.411766579Z 19:01:55 31 INFO  Monitor assigned new state "stop_replication"
2026-07-02T17:01:55.411768923Z 19:01:55 31 INFO  FSM transition from "prepare_promotion" to "stop_replication": Prevent against split-brain situations.
2026-07-02T17:01:55.411770961Z 19:01:55 31 INFO  Prevent writes to the promoted standby while the primary is not demoted yet, by making the service incompatible with target_session_attrs = read-write
2026-07-02T17:01:55.411771535Z 19:01:55 31 INFO  Setting default_transaction_read_only to on
2026-07-02T17:01:55.466877649Z 19:01:55 31 INFO  Reloading Postgres configuration and HBA rules
2026-07-02T17:01:55.521186885Z 19:01:55 31 INFO  Promoting postgres
2026-07-02T17:01:55.526511745Z 19:01:55 31 INFO  Waiting for postgres to promote
2026-07-02T17:01:56.536648755Z 19:01:56 31 INFO  Waiting for postgres to promote
2026-07-02T17:01:57.543068210Z 19:01:57 31 INFO  Waiting for postgres to promote
2026-07-02T17:01:58.548474084Z 19:01:58 31 INFO  Waiting for postgres to promote
2026-07-02T17:01:59.553870462Z 19:01:59 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:00.558081096Z 19:02:00 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:01.562588150Z 19:02:01 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:02.566552750Z 19:02:02 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:03.571251224Z 19:02:03 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:04.574953637Z 19:02:04 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:05.578399265Z 19:02:05 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:06.582321095Z 19:02:06 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:07.586311831Z 19:02:07 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:08.590055482Z 19:02:08 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:09.594327423Z 19:02:09 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:10.598487427Z 19:02:10 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:11.602838788Z 19:02:11 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:12.607445163Z 19:02:12 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:13.612117682Z 19:02:13 31 INFO  Waiting for postgres to promote
2026-07-02T17:02:14.684769020Z 19:02:14 31 INFO  Cleaning-up Postgres replication settings
2026-07-02T17:02:14.690602692Z 19:02:14 31 INFO  Disabling synchronous replication
2026-07-02T17:02:14.706030136Z 19:02:14 31 INFO  Reloading Postgres configuration and HBA rules
2026-07-02T17:02:14.745482875Z 19:02:14 31 INFO  Transition complete: current state is now "stop_replication"
2026-07-02T17:02:14.820683321Z 19:02:14 31 INFO  Monitor assigned new state "wait_primary"
2026-07-02T17:02:14.820686472Z 19:02:14 31 INFO  FSM transition from "stop_replication" to "wait_primary": Confirmed promotion with the monitor
2026-07-02T17:02:14.820689509Z 19:02:14 31 INFO  Setting default_transaction_read_only to off
2026-07-02T17:02:14.832036332Z 19:02:14 31 INFO  Reloading Postgres configuration and HBA rules
2026-07-02T17:02:14.852748036Z 19:02:14 31 INFO  Fetched current list of 1 other nodes from the monitor to update HBA rules, including 1 changes.
2026-07-02T17:02:14.853237488Z 19:02:14 31 INFO  Checking for HBA rules for node 1 "db001-1" (db001-1:5432)
2026-07-02T17:02:14.853239083Z 19:02:14 31 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl replication "pgautofailover_replicator" db001-1 trust
2026-07-02T17:02:14.853239526Z 19:02:14 31 WARN  Skipping HBA edits (per --skip-pg-hba) for rule: hostssl "postgres" "pgautofailover_replicator" db001-1 trust
2026-07-02T17:02:14.853241174Z 19:02:14 31 INFO  Transition complete: current state is now "wait_primary"
2026-07-02T17:02:36.712203623Z 19:02:36 31 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup
2026-07-02T17:02:44.494223976Z 19:02:44 31 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup

// short connection issue to monitor

2026-07-02T17:03:48.370127604Z 19:03:48 31 ERROR Connection to database failed: connection to server at "db001-monitor" (10.41.1.75), port 5432 failed: timeout expired
2026-07-02T17:03:48.370136678Z 19:03:48 31 ERROR connection to server at "db001-monitor" (10.41.1.193), port 5432 failed: timeout expired
2026-07-02T17:03:48.370137921Z 19:03:48 31 ERROR Failed to connect to monitor database at "postgres://autoctl_node@db001-monitor:5432/pg_auto_failover?sslmode=require", see above for details
2026-07-02T17:03:48.370192102Z 19:03:48 31 ERROR Failed to get the current version for extension "pgautofailover", see previous lines for details.
2026-07-02T17:03:48.370194277Z 19:03:48 31 ERROR Failed to get the goal state from the monitor

// reconnect

2026-07-02T17:03:55.489430221Z 19:03:55 31 INFO  Successfully got the goal state from the monitor
2026-07-02T17:10:07.698616868Z 19:10:07 31 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup
2026-07-02T17:54:35.617698269Z 19:54:35 31 INFO  New state for node 1 "db001-1" (db001-1:5432): primary ➜ catchingup

// db001-1 was manually removed from the cluster

2026-07-03T10:23:43.696685745Z 12:23:43 31 INFO  Monitor assigned new state "single"
```

I guess it is difficult to reconstruct what really happend given that some crucial logs are missing from node 1.

**Comment by dimitri** [comment]
Thanks for digging these up.

The container log is the useful part: a new `pg_autoctl` container for node1 started at 19:01:22, failed every monitor connection attempt, and was gone by 19:01:47 -- before the monitor even marked node1 unhealthy at 19:01:55. So node1's process was cycling right at the boundary of the incident, consistent with a stale/orphaned restart, but that specific container died before it could report anything, so it isn't the process that reported `primary` at 19:02:36.

One correction: the Postgres log you attached as node1's is almost certainly node151/db001-2's -- its promote-to-ready sequence (19:01:55 to 19:02:13-14) lines up to the second with db001-2's own pg_autoctl log in the same comment. So it doesn't tell us anything about node1 itself.

Net: still consistent with a second/orphaned process reporting a stale on-disk state, but not confirmed -- we'd need that later process's own logs, which don't seem to exist. Do you have any other clues from that window we haven't looked at yet?

**Comment by jnehlmeier** [comment]
> One correction: the Postgres log you attached as node1's is almost certainly node151/db001-2's -- its promote-to-ready sequence (19:01:55 to 19:02:13-14) lines up to the second with db001-2's own pg_autoctl log in the same comment. So it doesn't tell us anything about node1 itself.

I will verify tomorrow but I am pretty sure I looked it up for `db001-1` as I had no `pg_autoctl` logs available. 

But of course we recreated `db001-1` as secondary in the meantime and it did a basebackup of `db001-2` as part of `pg_autoclt` configuring the `secondary`. Does it also copy the log files because they live in `/var/lib/postgres/node/logs/*.log`? If yes, then we won't have any logs of `db001-1` anymore and the log above is indeed the one of `db001-2`.

**Comment by jnehlmeier** [comment]
Yeah log file of that day for both nodes are identical so it has been copied from `db001-2` to `db001-1` after we recreated `db001-1`.
